### PR TITLE
update string equality tests for h5py 3.1 compatibility

### PR DESCRIPTION
**Pull request by dwpaley**
h5py version 3 updated the rules for strings: https://docs.h5py.org/en/stable/strings.html. Some attributes that were previously returned as regular strings are now returned as bytes, which breaks tests like `if obj.handle["depends_on"][()] == ".":`. This PR converts both sides of all string equality tests in `format/nexus.py` to bytes using `numpy.string_`.

### CODE REVIEW

**Comment by graeme-winter** [comment]
@dwpaley strangely enough, while I recognise the changes you are making here, I did not need them for h5py 3.1 to work correctly for me (been using this already)

Much 🤔

In no sense is this a comment on the change set which I think you will find mirrored already on several branches...

**Comment by graeme-winter** [comment]
Asked @noemifrisina @benjaminhwilliams  to take a look as they have both been looking at this code closely lately

**Comment by phyy-nx** [comment]
Could have something to do with reading files in different versions of h5py than they were written. Similar to issues we had when we were doing the python 3 conversion.￼

File: format/nexus.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def visitor(name, obj):
                 numpy.string_("NXsubentry"),
             ]:
                 if "definition" in obj:
-                    if obj["definition"][()] == numpy.string_("NXmx"):
+                    if numpy.string_(obj["definition"][()]) == numpy.string_("NXmx"):
```
Comment (Anthchirp) [inline format/nexus.py:114]: If we _know_ that these strings are returned as bytes wouldn't
```suggestion
                    if obj["definition"][()] == b"NXmx":
```
be more natural?

Comment (Anthchirp) [reply format/nexus.py:114]: alternatively:
```suggestion
                    if obj["definition"][()] in ("NXmx", b"NXmx"):
```
if they can be both. Which would lend itself to be wrapped in a helper function such as
```python
def _h5match(obj, string):
    return obj[()] in (string, string.encode("utf-8"))

(...)
    if _h5match(obj["definition"], "NXmx"):
    (...)
```

**Comment by ndevenish** [comment]
By my reading
> Variable-length strings in attributes are read as str objects. These are decoded as UTF-8 with surrogate escaping for unrecognised bytes.
Implies that _all_ attributes are read as str objects now, and the only time `numpy` arrays are returned is for fixed-length datasets? So moving almost everything to plain str might be better, unless we want to support older `h5py` versions

**Comment by ndevenish** [comment]
> Implies that _all_ attributes are read as str objects now, and the only time `numpy` arrays are returned is for fixed-length datasets? So moving almost everything to plain str might be better, unless we want to support older `h5py` versions

Okay, this was me being hopelessly optimistic. It looks like, now, attribute strings _can be_ `str`, but can also be `numpy.string_` (which I assume is what they mean by `bytes`) - depending on how the attribute was written (fixed length or not). Rules appear to be summarized in https://github.com/h5py/h5py/issues/1338#issue-496665721. Additionally, it looks like `attrs` keys can be _both_ e.g. both `"x" in node.attrs` and `b"x" in node.attrs` work. 

So it'd have to be this, or the latter of Markus' suggestions (although maybe another variant - say a `h5str(something_out_of_h5py)` that means that only the left side needs to be converted).

**Comment by ndevenish** [comment]
I've got a prototype of this in https://github.com/cctbx/dxtbx/compare/nexus_h5py3...ndevenish:nexus_h5py3, which also seems to work

**Comment by ndevenish** [comment]
Since tests are breaking I've pulled this in now. Thanks!

**Pull request merged**